src/System.Private.CoreLib/shared/System/Array.cs

```diff
@@ -410,11 +402,11 @@ int IStructuralEquatable.GetHashCode(IEqualityComparer comparer)
         // negative result to produce the index of the first element (if any) that
         // is larger than the given search value.
         // 
-        public static int BinarySearch(Array array, object value)
+        public static int BinarySearch(Array array, object? value)
         {
             if (array == null)
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
-            return BinarySearch(array, array.GetLowerBound(0), array.Length, value, null);
+            return BinarySearch(array!, array!.GetLowerBound(0), array.Length, value, null); // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
```

Comment:
I'm surprised you had to add 2 `!` is that expected? Is it because it is on the same context?

Reply:
If the order were switched, I expect only one would be necessary, just on the first.  But the first one isn't actually dereferenced before the second one is, so the first is effectively casting an `object?` to an `object`, which deserves a warning but itself wouldn't actually cause null to be dereferenced, and the second actually may potentially dereference a null, as far as the compiler is concerned.  Both of these should become unnecessary once we can annotate ThrowHelper.

Reply:
Yeah, once we annotate throw helpers a lot of ! Will go away. Just wanted to understand why in this cases we need 2 but in other only 1. Thanks for explaining!